// @graphql-eslint/eslint-plugin/esm/configs/operations-all.js
import {
  __commonJS
} from "../chunk-BMTV3EA2.js";
var require_operations_all = __commonJS({
  "src/configs/operations-all.ts"(exports, module) {
    module.exports = {
      extends: "./configs/operations-recommended",
      rules: {
        "@graphql-eslint/alphabetize": [
          "error",
          {
            selections: ["OperationDefinition", "FragmentDefinition"],
            variables: ["OperationDefinition"],
            arguments: ["Field", "Directive"]
          }
        ],
        "@graphql-eslint/lone-executable-definition": "error",
        "@graphql-eslint/match-document-filename": [
          "error",
          {
            query: "kebab-case",
            mutation: "kebab-case",
            subscription: "kebab-case",
            fragment: "kebab-case"
          }
        ],
        "@graphql-eslint/no-one-place-fragments": "error",
        "@graphql-eslint/unique-fragment-name": "error",
        "@graphql-eslint/unique-operation-name": "error"
      }
    };
  }
});
export default require_operations_all();
